near-plugins-derive/src/upgradable.rs

```diff
@@ -49,9 +89,84 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
+                            staging_timestamp
+                        )
+                        .as_str(),
+                    );
+                }
+
                 near_sdk::Promise::new(near_sdk::env::current_account_id())
                     .deploy_contract(self.up_staged_code().unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: No staged code")))
             }
+
+            fn up_get_staging_timestamp(&self) -> Option<near_sdk::Timestamp> {
+                near_sdk::env::storage_read(self.up_staging_timestamp_storage_key().as_ref()).map(|staging_timestamp_bytes| {
+                    u64::from_be_bytes(staging_timestamp_bytes.try_into().unwrap_or_else(|_|
+                        near_sdk::env::panic_str("Upgradable: Invalid u64 timestamp format"))
+                    )
+                })
+            }
+
+            fn up_get_staging_duration(&self) -> Option<near_sdk::Duration> {
+                near_sdk::env::storage_read(self.up_staging_duration_storage_key().as_ref()).map(|staging_duration_bytes| {
+                    u64::from_be_bytes(staging_duration_bytes.try_into().unwrap_or_else(|_|
+                        near_sdk::env::panic_str("Upgradable: Invalid u64 Duration format"))
+                    )
+                })
+            }
+
+            #[#cratename::only(owner)]
+            fn up_init_staging_duration(&self, staging_duration: near_sdk::Duration) {
+                near_sdk::require!(self.up_get_staging_duration().is_none(), "Upgradable: staging duration was already initialized");
+                near_sdk::env::storage_write(self.up_staging_duration_storage_key().as_ref(), &staging_duration.to_be_bytes());
+            }
+
+            #[#cratename::only(owner)]
+            fn up_stage_update_staging_duration(&self, staging_duration: near_sdk::Duration) {
+                let staging_duration_timestamp = near_sdk::env::block_timestamp() + self.up_get_staging_duration().unwrap_or(0);
+                near_sdk::env::storage_write(self.up_update_staging_duration_storage_key().as_ref(), &staging_duration.to_be_bytes());
+                near_sdk::env::storage_write(self.up_update_staging_duration_timestamp_storage_key().as_ref(), &staging_duration_timestamp.to_be_bytes());
+            }
+
+            #[#cratename::only(owner)]
+            fn up_apply_update_staging_duration(&self) {
+                let staging_timestamp = self.up_get_update_staging_duration_timestamp()
+                    .unwrap_or_else(|| ::near_sdk::env::panic_str("Upgradable: No staged update"));
+
+                if staging_timestamp < near_sdk::env::block_timestamp() {
+                    near_sdk::env::panic_str(
+                        format!(
+                            "Upgradable: Update duration too early: staging ends on {}",
+                            staging_timestamp
+                        )
+                        .as_str(),
+                    );
+                }
```

Comment:
I feel like having a separate duration for updating the staging duration adds a lot of complexity. That's not mentioned in the NEP123, so just wondering if it's needed?

If the delay duration for deployments should be changeable, maybe it can be simplified by not letting the update affect a staged code upgrade?